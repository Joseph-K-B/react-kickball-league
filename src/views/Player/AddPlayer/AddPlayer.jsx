import { useState } from "react"
import { useHistory } from "react-router-dom"
import { createPlayer } from "../../../services/players";
import PlayerForm from "../../../components/PlayerForm";

function AddPlayer() {
    const [name, setName] = useState('');
    const [position, setPosition]= useState('');
    const [teamId, setTeamId] = useState(null);
    const history = useHistory();

    
    

    const handleAddPlayerSubmit = async (e) => {
        e.preventDefault();
        console.log('TEAM ID', teamId);
        const res = await createPlayer({name, position, teamId})
        history.push(`/players/${res[0].id}`)
    }

    // console.log(teamId);

    return(
        <>
        <fieldset>
            <legend>Add Player</legend>
            <PlayerForm 
                handleSubmit={handleAddPlayerSubmit}
                name={name}
                position={position}
                teamId={teamId}
                setName={setName}
                setPosition={setPosition}
                setTeamId={setTeamId}
            />
        </fieldset>
        </>
    )
}

export default AddPlayer;