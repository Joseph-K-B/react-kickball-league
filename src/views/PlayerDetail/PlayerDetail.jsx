import { useEffect, useState } from "react"
import { Link, useParams } from "react-router-dom"
import { getPlayerById } from "../../services/players"
import { getTeams } from "../../services/teams";

function PlayerDetail() {
    const{ id } = useParams()
    const[player, setPlayer] = useState(null);
    const [teams, setTeams] = useState([]);

    useEffect(() => {
        getPlayerById(id)
        .then((res) => setPlayer(res))
    }, [id])

    if(!player) return <h2>Loading player info...</h2>

    return(
        <>
        <h1>{player.name}</h1>
        <Link className='link' to={`/teams/${player.teams.id}`}>
            <h3>{player.teams.name}</h3>
        </Link>
        <h3>{player.teams.city}, {player.teams.state}</h3>        
        <h3>{player.position}</h3>
        <Link className='link' to='/players'><p>Player Directory</p></Link>
        </>
        
    )
}
export default PlayerDetail