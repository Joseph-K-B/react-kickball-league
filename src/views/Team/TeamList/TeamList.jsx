import {useEffect, useState} from 'react'
import { Link } from 'react-router-dom';
import { deleteTeamById, getTeams } from '../../../services/teams';

function TeamList() {
    const [loading, setLoading] = useState(true);
    const [teams, setTeams] = useState([]);

    const reloadTeams = async () => {
        setLoading(true);
        const res = await getTeams();
        setTeams(res);
        setLoading(false);
    }

    useEffect(() => {
        getTeams()
        .then((res) => setTeams(res))
        .finally(() => setLoading(false))
    }, [])

    const handleDelete = async ({ id, name }) => {
        // eslint-disable-next-line no-restricted-globals
        const confirmDelete = confirm(`Would you like to remove ${name} from this seasons roster?`);

        if(confirmDelete) {
            await deleteTeamById(id);
            await reloadTeams();
        }
    };

    if(loading) return <h1>Loading teams...</h1>
    
    return (
        <>
        <h1>Team List</h1>
        <ul>
            {teams.map((team) => {
                return (
                    <li key={team.id}>
                        <Link className='link' to={`/teams/${team.id}`}>
                        <p>{team.name}</p>
                        </Link>
                        <button
                            type="button"
                            aria-label={`Delete ${team.name}`}
                            onClick={() => handleDelete({ id: team.id, name: team.name})}
                        >
                            Delete
                        </button>
                        <Link to={`/teams/${team.id}/update`}>
                        <button
                            type="button"
                            aria-label={`Update ${team.name}`}
                        >
                            Update
                        </button>
                        </Link>
                    </li>
                    )
                })}
            </ul>            
            <Link className='link' to='/teams/new'><p>Add Team</p></Link>
            </>
    )
}

export default TeamList