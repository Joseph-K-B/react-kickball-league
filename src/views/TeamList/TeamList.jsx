import {useEffect, useState} from 'react'
import { Link } from 'react-router-dom';
import { getTeams } from '../../services/teams';

function TeamList() {
    const [loading, setLoading] = useState(true);
    const [teams, setTeams] = useState([]);

    useEffect(() => {
        getTeams()
        .then((res) => setTeams(res))
        .finally(() => setLoading(false))
    }, [])

    if(loading) return <h1>Loading teams...</h1>
    
    return (
        <>
        <h1>Team List</h1>
        <ul>
            {teams.map((team) => (
                <li key={team.id}>
                    <Link className='link' to={`/teams/${team.id}`}>
                    <p>{team.name}</p>
                    </Link>
                </li>
            ))}
        </ul>
        <Link className='link' to='/teams/new'><p>Add Team</p></Link>
        </>
    )
}

export default TeamList