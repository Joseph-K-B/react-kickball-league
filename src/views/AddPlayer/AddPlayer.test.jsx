import { render, screen } from '@testing-library/react';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import {Router, Route } from 'react-router-dom';
import { createMemoryHistory } from "history";
import userEvent from '@testing-library/user-event';
import AddPlayer from './AddPlayer';
import PlayerDetail from '../PlayerDetail/PlayerDetail';
import TeamList from '../TeamList/TeamList';
import { act } from 'react-dom/test-utils';

const mockPlayer = {
    id: 12,
    created_at: '2021-12-08T16:09:33.408898+00:00',
    name: 'player name',
    position: 'benchwarmer',
    teamId: 1
}

const mockTeam = {
    id: 1,
    created_at: '2021-12-08T16:09:33.408898+00:00',
    name: 'name',
    city: 'city',
    state: 'state',
    players: []
}

const server = setupServer(
    rest.get('https://vrmauzdnhcbzknrntyjm.supabase.co/rest/v1/teams', 
    (req, res, ctx) => {
        return res(ctx.json(mockTeam));
    }),
    rest.post('https://vrmauzdnhcbzknrntyjm.supabase.co/rest/v1/teams', 
    (req, res, ctx) => {
        return res (
            ctx.json([mockTeam]));
    }),
    rest.get('https://vrmauzdnhcbzknrntyjm.supabase.co/rest/v1/players', 
    (req, res, ctx) => {
        return res(ctx.json(mockPlayer));
    }),
    rest.post('https://vrmauzdnhcbzknrntyjm.supabase.co/rest/v1/players', 
    (req, res, ctx) => {
        return res (
            ctx.json([mockPlayer]));
    })
);

beforeAll(() => {
    server.listen();
});

afterAll(() => {
    server.close();
});

it('renders form to allow admin to add player and redirects to their details', async () => {
    const history = createMemoryHistory();
    history.push('/players/new');

    const {container} = render (
        <Router history={history}>
            {/* <Route exact path = '/teams' component={TeamList}/> */}
            <Route exact path = '/players/new'>
                <AddPlayer />
            </Route>
            <Route exact path = '/players/:id' component={PlayerDetail} />
        </Router>
    );

    screen.getByText('Add Player');

    const nameField = screen.getByLabelText(/Name:/);
    const positionField = screen.getByLabelText(/Position:/);
    const teamField = await screen.findByLabelText(/Team:/);
    // const optionChoice = await screen.findByRole('combobox');
    const submitBtn = screen.getByRole('button', {name: 'edit player'});

    act(() => {
        userEvent.type(nameField, 'New Team');
        userEvent.type(positionField, 'cheerleader');
        userEvent.selectOptions(screen.getByRole('combobox'), [mockTeam.id]);
        userEvent.click(submitBtn);
    })

    expect(screen.getByRole('option', {name: mockTeam.name}).ariaSelected).toBe(true)
    expect(container).toMatchSnapshot();
    await screen.findByText('player name');
});