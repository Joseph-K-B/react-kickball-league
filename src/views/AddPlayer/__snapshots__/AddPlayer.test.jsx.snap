// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders form to allow admin to add player and redirects to their details 1`] = `
<div>
  <fieldset>
    <legend>
      Add Player
    </legend>
    <form>
      <label
        for="name"
      >
        Name:
      </label>
      <input
        id="name"
        type="text"
        value="New Team"
      />
      <label
        for="position"
      >
        Position:
      </label>
      <input
        id="position"
        type="text"
        value="cheerleader"
      />
      <label
        for="teams"
      >
        Team:
      </label>
      <select
        id="teams"
      />
      <button
        aria-label="edit player"
        type="submit"
      >
        Submit
      </button>
    </form>
  </fieldset>
  <h2>
    Loading player info...
  </h2>
</div>
`;
